More instructions and small changes

jovobe opened the pull request:
- Added a bunch of new instructions
- Panics now pause the debugger instead of crashing it
- Fixed WRAM size
- Disabled trace/debug/info and warn in release mode

jovobe committed:
Add implementation of `Byte::rotate_left/rotate_right`

jovobe committed:
Add implementation for `RLC` and `RRC` instructions

jovobe committed:
Add implementation of `Byte::rotate_right_through_carry`

jovobe committed:
Add implementation for `RL` and `RR` instructions

jovobe committed:
Add implementation of `Byte::shift_left/shift_left`

jovobe committed:
Add implementation for `SLA` and `SRL` instructions

jovobe committed:
Add implementation of `Byte::arithmetic_shift_right`

jovobe committed:
Add implementation of `Byte::swap_nybbles`

jovobe committed:
Add implementation for `SRA` and `SWAP` instructions

jovobe committed:
Add implementation of `Not` for `Byte`

jovobe committed:
Add implementation for `BIT`, `SET` and `RES` instructions

jovobe committed:
Pause the emulator on panic!

This is very useful to see where the panic happened.

jovobe committed:
Add OAM getter/setter to `Ppu` and adjust memory mapping to it

jovobe committed:
Fix wrong WRAM size in `Machine::new`

jovobe committed:
Add convenience method `Machine::pop`

jovobe committed:
Make use of `Machine::load_hl/store_hl/pop` in `step.rs`

jovobe committed:
Add implementation for further instructions

jovobe committed:
Add implementation of `Word::add_with_carries`

jovobe committed:
Add implementation for further instructions

jovobe committed:
Disable trace/debug/info and warn in release mode

LukasKalbertodt commented on core/src/machine/mm.rs:
```diff
@@ -44,8 +48,12 @@ impl Machine {
+            0xFEA0..0xFF00 => {
+                // On DMG writes to this are ignored
+                // TODO: Add correct CGB behavior
+                trace!("Wrote to {:?} which is in not writable range: 0xFEA0..0xFF00!", addr);
```
`{}` as placeholder works, too. `addr` implements `Display`.

LukasKalbertodt commented on core/src/machine/ppu.rs:
```diff
@@ -121,6 +121,36 @@ impl Ppu {
+    pub(crate) fn load_oam_byte(&self, addr: Word) -> Byte {
+        match self.phase() {
+            Phase::PixelTransfer => Byte::new(0xff),
+            Phase::OamSearch => Byte::new(0xff),
```
Maybe rather `Phase::PixelTransfer | Phase::OamSearch => Byte::new(0xff)`?

jovobe replied:
I thought about this too, but think the current solution is more explicit and better readable.

LukasKalbertodt replied:
I disagree :D I think it's just duplicate code (`Byte::new(0xff)`) in your version.

But I won't fight over this, so we can keep it ;-)

LukasKalbertodt commented on core/src/machine/ppu.rs:
```diff
@@ -121,6 +121,36 @@ impl Ppu {
+        match self.phase() {
+            Phase::PixelTransfer => Byte::new(0xff),
+            Phase::OamSearch => Byte::new(0xff),
+            _ => self.vram[addr - 0xFE00],
```
Oopsie daisy, I guess this should be `self.oam`

LukasKalbertodt commented on core/src/machine/ppu.rs:
```diff
@@ -121,6 +121,36 @@ impl Ppu {
+        match self.phase() {
+            Phase::PixelTransfer => {},
+            Phase::OamSearch => {},
+            _ => self.vram[addr - 0xFE00] = byte,
```
Same as above.

LukasKalbertodt commented on desktop/src/main.rs:
```diff
@@ -124,8 +138,13 @@ fn run() -> Result<(), Error> {
     log::set_max_level(log::LevelFilter::Trace);
 
+    #[cfg(not(debug_assertions))]
+    log::set_max_level(log::LevelFilter::Error);
```
I'd say `Warn` is a better level for release mode.

jovobe replied:
I think that's a matter of taste. I try to worry the user as little as possible. And warnings can make the user feel uncomfortable using the emulator (which is not necessary for things we pump out warnings at the moment (e.g. writing to the cartridge)).

LukasKalbertodt replied:
Ah, yes, true. Wasn't really thinking about "the user" yet, but rather about us dev who want to run the program with optimizations.

In any case, for your information: in the Rust world the logger implementation `env_logger` is basically the standard. It is configurable at runtime via the `RUST_LOG` environment variable. So the user can override it at will (`Trace` usually completely disabled in release mode though, just due to speed). But we can worry about this polishing stuff later. Also #26 is obviously related.

jovobe replied:
Ah, good to know!

LukasKalbertodt commented on core/src/machine/step.rs:
```diff
@@ -117,6 +117,17 @@ impl Machine {
         }
 
+        /// This is a template macro for all ADD HL, w instructions (where `w` should
+        /// be a [`Byte`]).
```
"Should be a `Word`", right?

LukasKalbertodt commented on core/src/machine/step.rs:
```diff
@@ -182,6 +193,85 @@ impl Machine {
+        macro_rules! sla {
+            ($x:expr) => {{
+                let carry = $x.shift_left();
+                let zero = self.cpu.c == Byte::zero();
```
Why `self.cpu.c`? I'm fairly sure this should be `$x == Byte::zero()`. This error is all over the place. I won't comment on every position: just check the macros before and after this one.

LukasKalbertodt reviewed (changes requested):
I don't have enough brain left today to finish this review. I "only" need to finish reviewing `step.rs`. Will do that sometime tomorrow.

LukasKalbertodt commented on core/src/machine/step.rs:
```diff
@@ -414,6 +530,40 @@ impl Machine {
+            opcode!("RST 38H") => {
+                self.push(self.cpu.pc);
+                self.cpu.pc = Word::new(0x38);
+            }
```
I guess a `call!` macro would be handy here. This would work for all RST and all CALL instructions.

LukasKalbertodt commented on core/src/machine/step.rs:
```diff
@@ -433,12 +583,59 @@ impl Machine {
 
+            // ========== JP ==========
+            opcode!("JP a16") => self.cpu.pc = arg_word,
+            opcode!("JP (HL)") => self.cpu.pc = self.cpu.hl(),
```
Mh I think the mnemonic is misleading here (not the fault of this PR). I would change it to `JP HL`. It not called `JP (a16)` either. And [this page](http://gbdev.gg8.se/wiki/articles/CPU_Instruction_Set) also calls it `JP HL`. Maybe you could change that in this PR still (including `instr.rs`, of course).

LukasKalbertodt reviewed (changes requested):
Ok, finally finished review. Apart from the comments I left, this looks good. Cant' wait to test it out :grin:

jovobe committed:
Rename instruction `JP (HL)` into `JP HL`

The name `JP (HL)` is misleading, because the `()` notation is normally
used to indicate a value, read from the memory which is not the case
in this instruction.

jovobe committed:
Add convenience macro `call!` for CALL-like instructions

jovobe committed:
Fix copy-paste bug in RL, RR, SLA, SRL, SRA and SWAP instructions

jovobe committed:
Fix copy-paste bug in the memory mapping of the `Ppu`

jovobe committed:
Fix copy-paste bug in comment of `add_hl!`

jovobe committed:
Switch debug formatter to display formatter for some `Word` output
